shared/actions/wallets.js

```diff
@@ -41,12 +44,64 @@ const loadPaymentsSuccess = (res: RPCTypes.PaymentsPageLocal, action: WalletsGen
   )
 }
 
+const linkExistingAccount = (state: TypedState, action: WalletsGen.LinkExistingAccountPayload) => {
+  const {name, secretKey} = action.payload
+  return RPCTypes.localLinkNewWalletAccountLocalRpcPromise(
+    {
+      name,
+      secretKey: secretKey.stringValue(),
+    },
+    Constants.linkExistingWaitingKey
+  )
+    .then(accountID => [
+      WalletsGen.createLoadAccounts(),
+      WalletsGen.createSelectAccount({accountID: Types.stringToAccountID(accountID), show: true}),
+      WalletsGen.createLinkExistingAccountError({error: '', name, secretKey}),
+    ])
+    .catch(err => WalletsGen.createLinkExistingAccountError({error: err.desc, name, secretKey}))
+}
+
+const validateAccountName = (state: TypedState, action: WalletsGen.ValidateAccountNamePayload) => {
+  const {name} = action.payload
+  return RPCTypes.localValidateAccountNameLocalRpcPromise({name})
+    .then(() => WalletsGen.createValidateAccountNameError({error: '', name}))
+    .catch(err => WalletsGen.createValidateAccountNameError({error: err.desc, name}))
+}
+
+const validateSecretKey = (state: TypedState, action: WalletsGen.ValidateSecretKeyPayload) => {
+  const {secretKey} = action.payload
+  return RPCTypes.localValidateSecretKeyLocalRpcPromise({secretKey: secretKey.stringValue()})
+    .then(() => WalletsGen.createValidateSecretKeyError({error: '', secretKey}))
+    .catch(err => WalletsGen.createValidateSecretKeyError({error: err.desc, secretKey}))
+}
+
+const navigateToAccount = (action: WalletsGen.SelectAccountPayload) => {
+  const {show} = action.payload
+  if (!show) {
+    return
+  }
+  return Saga.put(Route.navigateTo([{props: {}, selected: walletsTab}, {props: {}, selected: null}]))
+}
+
 function* walletsSaga(): Saga.SagaGenerator<any, any> {
   yield Saga.safeTakeEveryPure(WalletsGen.loadAccounts, loadAccounts, loadAccountsSuccess)
   yield Saga.safeTakeEveryPure(WalletsGen.loadAssets, loadAssets, loadAssetsSuccess)
   yield Saga.safeTakeEveryPure(WalletsGen.loadPayments, loadPayments, loadPaymentsSuccess)
   yield Saga.safeTakeEveryPure(WalletsGen.selectAccount, loadAssets, loadAssetsSuccess)
   yield Saga.safeTakeEveryPure(WalletsGen.selectAccount, loadPayments, loadPaymentsSuccess)
+  yield Saga.safeTakeEveryPurePromise(
+    action => action.type === WalletsGen.linkExistingAccount && !action.error,
```

Comment:
i'm not necessarily against this form but it might be useful to just have the handler check for error, so as the acceptable types grows that check can be there instead of these patterns getting longer

Reply:
I started out doing that with `safeTakeEveryPure` and `actionCreatorsWithResult`, etc. The issue there was that `actionCreatorsWithResult/Error` always ran and made a `createLinkExistingAccountError` (either to set or clear), which resulted in an infinite loop if I forgot to check for the error in those handlers too. With this form I can change it to check in the one handler :+1: